rust/src/qfunction.rs

```diff
@@ -0,0 +1,532 @@
+///   v.iter_mut()
+///     .zip(u.iter().zip(weights.iter()))
+///     .for_each(|(v, (u, w))| *v = u * w);
+///
+///   // Return clean error code
+///   0
+/// };
+///
+/// let mut qf = ceed.q_function_interior(1, Box::new(user_f));
+///
+/// qf.add_input("u", 1, EvalMode::Interp);
+/// qf.add_input("weights", 1, EvalMode::Weight);
+///
+/// qf.add_output("v", 1, EvalMode::Interp);
+///
+/// println!("{}", qf);
+/// ```
+impl fmt::Display for QFunction {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.qf_core.fmt(f)
+    }
+}
+
+/// View a QFunction by Name
+///
+/// ```
+/// # use libceed::prelude::*;
+/// # let ceed = libceed::Ceed::default_init();
+/// let qf = ceed.q_function_interior_by_name("Mass1DBuild");
+/// println!("{}", qf);
+/// ```
+impl fmt::Display for QFunctionByName {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.qf_core.fmt(f)
+    }
+}
+
+// -----------------------------------------------------------------------------
+// Core functionality
+// -----------------------------------------------------------------------------
+impl QFunctionCore {
+    // Common implementation
+    pub fn apply(&self, Q: i32, u: &[Vector], v: &[Vector]) {
+        let mut u_c = [std::ptr::null_mut(); crate::MAX_QFUNCTION_FIELDS];
+        for i in 0..std::cmp::min(crate::MAX_QFUNCTION_FIELDS, u.len()) {
+            u_c[i] = u[i].ptr;
+        }
+        let mut v_c = [std::ptr::null_mut(); crate::MAX_QFUNCTION_FIELDS];
+        for i in 0..std::cmp::min(crate::MAX_QFUNCTION_FIELDS, v.len()) {
+            v_c[i] = v[i].ptr;
+        }
+        unsafe {
+            bind_ceed::CeedQFunctionApply(self.ptr, Q, u_c.as_mut_ptr(), v_c.as_mut_ptr());
+        }
+    }
+}
+
+// -----------------------------------------------------------------------------
+// User QFunction Closure
+// -----------------------------------------------------------------------------
+pub type QFunctionUserClosure = dyn FnMut(usize, &[&[f64]], &mut [&mut [f64]]) -> i32;
+
+unsafe extern "C" fn trampoline(
+    ctx: *mut ::std::os::raw::c_void,
+    q: bind_ceed::CeedInt,
+    inputs: *const *const bind_ceed::CeedScalar,
+    outputs: *const *mut bind_ceed::CeedScalar,
+) -> ::std::os::raw::c_int {
+    let context = &mut *(ctx as *mut QFunction);
+    let trampoline_data = &context.trampoline_data;
+
+    // Inputs
+    let inputs_slice: &[*const bind_ceed::CeedScalar] =
+        std::slice::from_raw_parts(inputs, crate::MAX_QFUNCTION_FIELDS);
+    let mut inputs_vec: Vec<&[f64]> = Vec::new();
+    for i in 0..trampoline_data.number_inputs {
+        inputs_vec.push(&std::slice::from_raw_parts(
+            inputs_slice[i],
+            (trampoline_data.input_sizes[i] * q) as usize,
+        ) as &[f64]);
+    }
+
+    // Outputs
+    let outputs_slice: &[*mut bind_ceed::CeedScalar] =
+        std::slice::from_raw_parts(outputs, crate::MAX_QFUNCTION_FIELDS);
+    let mut outputs_vec: Vec<&mut [f64]> = Vec::new();
```

Comment:
Use `[&mut [f64], MAX_QFUNCTION_FIELDS]`?

Reply:
[arrayvec](https://docs.rs/arrayvec/0.5.2/arrayvec/struct.ArrayVec.html) gives the `Vec` interface for stack-allocated arrays, though a raw array also seems fine here.

Somewhat related to this review, I don't know how to provide interior mutability without a much more complicated type signature (e.g., using `RefCell`).
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b33c4708959797511baf72a8be6fc965

Reply:
Is it what you want?
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=112b28c2b8892bd520bc7725553eef44

Reply:
That will not allow assignment to a mutable variable, which seems desirable
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=112b28c2b8892bd520bc7725553eef44

Reply:
I thought this was what @jedbrown wanted

Reply:
Well, I think its good to allow
```
mut v = &mut outputs[0];
```
and disallow
```
outputs[0] = v
```
if we can, but I feel more strongly about allowing the first.

Reply:
I wanted the interior (array entry) mutability without ownership (i.e., slice, not array). But I suppose that as a function argument, it'd be okay to move the array, as in
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=38ba82d5e1bf110ea4fe5075eabf1fe2

Trouble is that the length of the array is now part of the function signature (maybe this is also a good thing). I'm not sure the best way to express it, but it'd be better if we could just make them all function arguments. It seems like macro should be able to determine the number of arguments and mutability of those arguments in order to do the transformation, but I think we could work out that refinement in a subsequent PR.

Reply:
And the move semantics produce good code (no copies, just a bounds check).
```
example::f:
        push    rax
        cmp     rsi, 2
        jb      .LBB6_2
        add     dword ptr [rdi + 4], 50
        pop     rax
        ret

...
        mov     rdi, qword ptr [rsp + 80]
        mov     rsi, qword ptr [rsp + 88]
        call    example::f
```
for
```rust
#[inline(never)]
fn f(a: [&mut [i32]; 2]) {
    a[0][1] += 50;
}

pub fn main() {
    let mut r0 = [1, 2];
    let mut r1 = [3, 4];
    let z: [&mut [i32]; 2] = [&mut r0, &mut r1];
    z[0][1] = 5;
    println!("z {:?}", z);
    f(z);         // z moved here, but r0 and r1 are still valid
    println!("z {:?}", [r0, r1])
}
```
https://gcc.godbolt.org/z/GoKzGx

Reply:
I think when I tried this before I had to set the array length to the max possible, which has us back at 
```
[&mut [f64], MAX_QFUNCTION_FIELDS]
```

Reply:
A good use case for const generics?
https://without.boats/blog/shipping-const-generics/

Reply:
I also think that we should be able to do with Rust macros (`qfunction!`) something with the same flavor as what @pazner has done for the Julia interface. DSL are also extremely easy to write with Rust which should help us write very elegantly `CeedOperator`. I agree that none of this needs to be in the first release of the Rust interface.